Arbitrary region support

ollybritton opened the pull request.

ollybritton committed:
ENH: allow non-rectangular regions through `region` argument

ollybritton committed:
API: switch to `region` as a list of callables

This means a larger class of regions are supported.

ollybritton committed:
ENH: add support for array api to points argument

ollybritton committed:
ENH: array API support for infinite limits

- Also added a test for when the region of integration has zero width in
  one dimension

ollybritton committed:
ENH: support array API when using `region` keyword

ollybritton committed:
TST: modify tests to be faster

- Reduce the bounds of one of the `test_scalar_output` tests so that it doesn't take as long
- Skip some high-dimensional Genz-Malik integrals when using `array_api_strict` since these were very slow (>3 minutes)

ollybritton committed:
DOC: docs for infinite limits, `region` and `CubatureResult`

- Remove old text in docs saying infinite limits were not allowed
- Add examples for infinite limits and `region`
- Don't refer to `CubatureResult` in the docs since this is not public
- Also add test for making sure `a` and `b` are nonempty
- Also add test limits of the form `(a, infty)` or `(-infty, b)`

ollybritton committed:
DOC: ignore line too long warning

ollybritton committed:
ENH: specify `points` in the original coordinates when using inf limits

- Still need to implement this for the function limits transform

izaid commented on scipy/integrate/_cubature.py:
```diff
@@ -347,3 +718,406 @@ def _process_subregion(data):
+    return xp.all(a <= point) and xp.all(point <= b)
+
+
+def _split_at_points(a, b, points):
```
I think this can be done simpler... Don't you just need to sort and check for infinite points?

I think this can be done actually much simpler... Would also be good to avoid the loops I played around with it, something like the following? (Needs to be adapted).

`xp.concatenate([xp.expand_dims(xp.asarray(a), axis = 0), xp.sort(points, axis = 0), xp.expand_dims(xp.asarray(b), axis = 0)], axis = 0)`

ollybritton committed:
ENH: specify `points` in the original coordinates when using func limits

ollybritton committed:
DOC: update `cubature` docstring to include info for `region`, inf limits

- Add more comments explaining the transformations themselves
- Remove `t_to_x` and rename `x_to_t` as `inv`

ollybritton committed:
DOC: further doc updates for clarity

- Switch order of notes and examples in docstring to match how it is
  rendered
- Use consistent reshaping syntax
- Make it more obvious that the not working example won't work

ollybritton committed:
TST: use xslow rather than skipping very slow test

ollybritton committed:
MAINT: clarify function limit transformation code

- Remove confusing `x_and_y` and `x_and_t` variable names
- Add more explanatory comments
- Bring `inv` and `_inner_limits` more in-line with one another since
  they do very similar things

ollybritton committed:
MAINT: rename `regions` to `patches` for clarity with `region` argument

ollybritton committed:
BUG: remove arguments `a`, `b` from property

ollybritton committed:
BUG: address array-api related CI failures

- Switch from `xp.asarray([a, b])` to `xp.stack([a, b])`
- Use `xp.prod` rather than `math.prod`
- Add test skip for JAX on tests related to transformations since JAX
  doesn't support `a[i] = ...` or `t[..., i] = ...`.
- Use `xp_size` rather than `.size`
- Use compatible namespace in tests so that `xp.repeat` is available

ollybritton committed:
TST: mark very slow tests (>5s) with xslow

ollybritton committed:
DOC: add missing `::` to prevent doc build warning

ollybritton committed:
TST: mark slow test and don't use xp.repeat

- Mark another test as slow since it was taking too long to run on
  windows
- Don't use `xp.repeat` since it is not available for PyTorch

ollybritton committed:
TST: correct `jax` to `jax.numpy` in `skip_xp_backends`
